### PR TITLE
advent-2016: timers in Go

### PR DESCRIPTION
**Pull request by LK4D4**
Article for Advent 2016 event about timers in Go. Timers include
time.Sleep, time.Timer and time.Ticker.

I hope it's okay to start new `advent-2016` directory. I marked the article as a draft for now, so it doesn't show on generated site.

ping @bketelsen @dgryski

### CODE REVIEW

**Comment by LK4D4** [comment]
Thanks @ScottMansfield for the review!
Maybe it's easier to read it at https://github.com/LK4D4/lk4d4.darth.io/blob/master/content/posts/timers.md btw

File: content/advent-2016/go-timers.md [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+		// timers.waitnote can be used to wakeup this goroutine with notewakeup.
+		timers.sleeping = true
+		noteclear(&timers.waitnote)
+		unlock(&timers.lock)
+		notetsleepg(&timers.waitnote, delta)
+	}
+}
+```
+There are two variables which I think deserve explanation: `rescheduling` and
+`sleeping`. They both indicate that goroutine was put to sleep, but different
```
Comment (nathany) [inline content/advent-2016/go-timers.md]: that **the** goroutine

File: content/advent-2016/go-timers.md [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+synchronization mechanisms are used, let's discuss them.
+
+- `sleeping` is set when all "current" timers are processed, but there is more
```
Comment (nathany) [inline content/advent-2016/go-timers.md]: there **are** more

File: content/advent-2016/go-timers.md [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+which we need to spawn in future. It uses OS-based synchronization, so it calls
+some OS syscalls to put to sleep and wake up goroutine and syscalls means it spawns
```
Comment (nathany) [inline content/advent-2016/go-timers.md]: wake up **the** goroutine

File: content/advent-2016/go-timers.md [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+structure and next functions for synchronization:
+	* [noteclear](https://github.com/golang/go/blob/2f6557233c5a5c311547144c34b4045640ff9f71/src/runtime/lock_futex.go#L125) -
+	resets note state.
+	* [notetsleepg](https://github.com/golang/go/blob/2f6557233c5a5c311547144c34b4045640ff9f71/src/runtime/lock_futex.go#L199) - 
+	puts goroutine to sleep until `notewakeup` is called or after some period
+	of time (in case of timers it's time until next timer). This func fills `timers.waitnote`
+	with "pointer to timer goroutine".
+	* [notewakeup](https://github.com/golang/go/blob/2f6557233c5a5c311547144c34b4045640ff9f71/src/runtime/lock_futex.go#L129) - 
+	wakes up goroutine which called `notetsleepg`.
+`notewakeup` might be called in `addtimerLocked` if new timer is "earlier" than
```
Comment (nathany) [inline content/advent-2016/go-timers.md]: if **the** new timer

Comment (nathany) [inline content/advent-2016/go-timers.md]: than **the** previous

File: content/advent-2016/go-timers.md [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+previous "earliest" timer.
+
+- `rescheduling` is set when there are no timers in our heap, so nothing to do.
+It uses go scheduler to put goroutine to sleep with function
```
Comment (nathany) [inline content/advent-2016/go-timers.md]: uses **the** go scheduler

Comment (nathany) [inline content/advent-2016/go-timers.md]: put **the** goroutine

File: content/advent-2016/go-timers.md [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+[goparkunlock](https://github.com/golang/go/blob/2f6557233c5a5c311547144c34b4045640ff9f71/src/runtime/proc.go#L264).
+Unlike `notetsleepg` it does not consume any OS resources, but also does not
+support "wakeup timeout" so it can't be used instead of `notetsleepg` in
```
Comment (nathany) [inline content/advent-2016/go-timers.md]: in **the** `sleeping`

File: content/advent-2016/go-timers.md [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+`sleeping` branch.
+[goready](https://github.com/golang/go/blob/2f6557233c5a5c311547144c34b4045640ff9f71/src/runtime/proc.go#L264)
```
Comment (nathany) [inline content/advent-2016/go-timers.md]: The **goready**

File: content/advent-2016/go-timers.md [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+function is used for waking up goroutine when new timer added with `addTimerLocked`.
```
Comment (nathany) [inline content/advent-2016/go-timers.md]: waking up **the** goroutine

Comment (nathany) [inline content/advent-2016/go-timers.md]: when **a** new timer **is** added

File: content/advent-2016/go-timers.md [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+
+## Conclusion
+
+We learned how Go timers work "under the hood" - runtime neither uses one goroutine per
```
Comment (nathany) [inline content/advent-2016/go-timers.md]: **the** runtime

File: content/advent-2016/go-timers.md [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+timer, nor timers are "free" to use. It's important to understand how things work
```
Comment (nathany) [inline content/advent-2016/go-timers.md]: nor **are** timers "free"

File: content/advent-2016/go-timers.md [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+to avoid premature optimizations. Also, we learned that it's quite easy to read
+runtime code and you shouldn't be afraid to so.
```
Comment (nathany) [inline content/advent-2016/go-timers.md]: afraid to **do** so

File: content/advent-2016/go-timers.md [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+I hope you enjoyed this reading and will share info to your fellow Gophers.
```
Comment (nathany) [inline content/advent-2016/go-timers.md]: 👍

File: content/advent-2016/go-timers.md [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+
+# How Do They Do It: Timers in Go
+
+This article covers internal implementation of timers in Go. Note that there are
+a lot of links to Go repo in this article, I recommend to follow them to understand
+material better.
+
+## Timers
+
+Timers in Go just do something after a period of time. The user interface for
```
Comment (LK4D4) [inline content/advent-2016/go-timers.md]: Idea was that there are also non-user timers, which users can't use. I think maybe just "Timers are located".

**Comment by LK4D4** [comment]
@nathany thanks for a thorough review. I hope to truly understand articles usage in English one day :)

**Pull request merged**